File: Platforms/Common/Reign_Video_D3D11_Component/Shaders/VertexShader.cpp
#include "pch.h"
#include "VertexShader.h"

namespace Reign_Video_D3D11_Component
{
	#pragma region Constructors
	VertexShaderErrors VertexShaderCom::Init(VideoCom^ video, ShaderModelCom^ shaderModel)
	{
		null();
		this->video = video;
		resources = shaderModel->resources;
		variableBuffer = shaderModel->variableBuffer;
		resourceKnownCount = shaderModel->resourceKnownCount;

		ID3D11VertexShader* shaderTEMP;
		#if WIN32
		ID3DBlob* codeBlob = shaderModel->code;
		if (FAILED(video->device->CreateVertexShader(codeBlob->GetBufferPointer(), codeBlob->GetBufferSize(), NULL, &shaderTEMP)))
		#else
		if (FAILED(video->device->CreateVertexShader(shaderModel->code, shaderModel->codeSize, NULL, &shaderTEMP)))
		#endif
		{
			return VertexShaderErrors::VertexShader;
		}
		shader = shaderTEMP;

		return VertexShaderErrors::None;
	}

	VertexShaderCom::~VertexShaderCom()
	{
		if (shader) shader->Release();
		null();
	}

	void VertexShaderCom::null()
	{
		shader = 0;
	}
	#pragma endregion

	#pragma region Methods
	void VertexShaderCom::Apply()
	{
		if (variableBuffer)
		{
			ID3D11Buffer* buffer = variableBuffer;
			video->deviceContext->VSSetConstantBuffers(0, 1, &buffer);
		}

		if (resourceKnownCount != 0 && resources)
		{
			video->deviceContext->VSSetShaderResources(0, resourceKnownCount, resources);
		}

		video->deviceContext->VSSetShader(shader, NULL, 0);
	}
	#pragma endregion
}

File: Platforms/Common/Reign.Video.D3D11/Shaders/VertexShader.cs
using System;
using Reign.Core;
using Reign.Video;
using Reign_Video_D3D11_Component;

namespace Reign.Video.D3D11
{
	public class VertexShader : ShaderModel
	{
		#region Properties
		internal VertexShaderCom vertexShaderCom;
		#endregion

		#region Constructors
		#if WINDOWS
		public VertexShader(Shader shader, string code, ShaderVersions shaderVersion)
		: base(shader, code, ShaderTypes.VS, shaderVersion)
		#else
		public VertexShader(Shader shader, byte[] code)
		: base(shader, code, ShaderTypes.VS)
		#endif
		{
			try
			{
				var video = shader.FindParentOrSelfWithException<Video>();
				vertexShaderCom = new VertexShaderCom();
				var error = vertexShaderCom.Init(video.com, com);

				if (error == VertexShaderErrors.VertexShader) Debug.ThrowError("VertexShader", "Failed to create VertexShader");
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (vertexShaderCom != null)
			{
				vertexShaderCom.Dispose();
				vertexShaderCom = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public override void Apply()
		{
			base.Apply();
			vertexShaderCom.Apply();
		}

		#if METRO || WP8
		public override int Variable(string name)
		{
			for (int i = 0; i != shader.vsVariableNames.Count; ++i)
			{
				if (shader.vsVariableNames[i] == name) return shader.vsVariableByteOffsets[i];
			}

			return -1;
		}

		public override int Resource(string name)
		{
			for (int i = 0; i != shader.vsResourceCount; ++i)
			{
				if (shader.vsResourceNames[i] == name) return shader.vsResourceIndices[i];
			}

			return -1;
		}
		#endif
		#endregion
	}
}
